cuqi/randomvariable/_randomvariable.py

```diff
@@ -0,0 +1,191 @@
+from __future__ import annotations
+from typing import List, Any
+from ._ast import RandomVariableNode
+from ._orderedset import OrderedSet
+import operator
+import cuqi
+
+
+class RandomVariable:
+    """ Random variable defined by a base distribution with potential algebraic operations applied to it.
+
+    The random variable class is still experimental.
+
+    The random variable can be viewed as a lazily evaluated array. This allows for the definition of
+    Bayesian models in a natural way.
+
+    Example
+    -------
+
+    Basic usage:
+
+    .. code-block:: python
+
+        from cuqi.distribution import Gaussian
+
+        x = Gaussian(0, 1)
+
+        print(x+1)  # returns a RandomVariable object with operation recorded
+
+    Directly creating random variable:
+   
+    .. code-block:: python
+
+        from cuqi.distribution import Gaussian
+
+        x = Gaussian(0, 1)
+
+        x_rv = x.as_random_variable()  # returns a RandomVariable object
+
+        print((x_rv+1)**2)  # returns a RandomVariable object with operation recorded
+
+    Defining Bayesian problem using algebra on distributions (creating random variables implicitly):
+
+    .. code-block:: python
+
+        from cuqi.testproblem import Deconvolution1D
+        from cuqi.distribution import Gaussian, Gamma, GMRF
+        from cuqi.problem import BayesianProblem
+        import numpy as np
+        A, y_obs, info = Deconvolution1D.get_components()
+
+        # Bayesian problem
+        d = Gamma(1, 1e-4)
+        s = Gamma(1, 1e-4)
+        x = GMRF(np.zeros(A.domain_dim), d)
+        y = Gaussian(A @ x, 1/s)
+
+        BP = BayesianProblem(y, x, s, d)
+        BP.set_data(y=y_obs)
+        BP.UQ()
+
+    """
+
+    def __init__(self, distributions: set, tree: RandomVariableNode = None):
+        """ Create random variable from distribution """
+        # Convert to list if single distribution
+        if not isinstance(distributions, OrderedSet):
+            distributions = OrderedSet([distributions])
```

Comment:
I was wondering what is the difference between the OrderedSet here and for example what we use before: "processed_kwargs = set() " in distributions.

Reply:
Ordered set allows the order to be maintained which is e.g. used in Distribution if one of the parameters is a RV and we use positional arguments to condition.